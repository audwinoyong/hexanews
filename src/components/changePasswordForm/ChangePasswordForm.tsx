/**
 * Display Change Password Form component for user to change their password
 */

import * as React from 'react'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import { Typography } from '@material-ui/core'
import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'

import * as actions from './actions'
import { accountString } from '../../constants/string'
import { User } from '../../domain/model/User'
import selector, { StateProps } from './selector'
import styles from './styles'

export type Props = {
  user: User,
  dispatch: Dispatch<any>,
  isChangingPassword: boolean,
} & StateProps

export interface ComponentState {
  email: string,
  currentPassword: string,
  newPassword: string,
  confirmNewPassword: string,
}

export class ChangePasswordForm extends React.Component<Props, ComponentState> {

  constructor(props: Props) {
    super(props)
    const { user } = this.props
    this.state = {
      email: user.email,
      currentPassword: '',
      newPassword: '',
      confirmNewPassword: '',
    }
  }

  componentDidMount() {
    // Custom rule validation to confirm new password matches
    ValidatorForm.addValidationRule('isNewPasswordMatch', (value: any) => {
      if (value !== this.state.newPassword) {
        return false
      }
      return true
    })
  }

  handleCurrentPasswordChange = (event: any) => {
    this.setState({ currentPassword: event.target.value })
  }

  handleNewPasswordChange = (event: any) => {
    this.setState({ newPassword: event.target.value })
  }

  handleConfirmNewPasswordChange = (event: any) => {
    this.setState({ confirmNewPassword: event.target.value })
  }

  disableChangePasswordButton() {
    const { currentPassword, newPassword, confirmNewPassword } = this.state
    return (currentPassword === '' || newPassword === '' || confirmNewPassword === '')
  }

  onChangePassword = async () => {
    const { dispatch } = this.props
    const { email, currentPassword, newPassword } = this.state

    await dispatch(actions.changePassword(email, currentPassword, newPassword))
  }

  public render() {
    const { isChangingPassword } = this.props
    const { currentPassword, newPassword, confirmNewPassword } = this.state
    return (
      <div style={styles.sectionContainer}>
        <ValidatorForm
          ref='changePasswordForm'
          onSubmit={() => this.onChangePassword()}
          instantValidate={false}
        >
          <Typography
            variant='headline'
            component='h2'
            color='textPrimary'
          >
            {accountString.changePassword}
          </Typography>
          <TextValidator
            label='Current Password'
            onChange={this.handleCurrentPasswordChange}
            name='currentPassword'
            style={styles.textField}
            margin='dense'
            type='password'
            value={currentPassword}
            helperText=' '
            validators={['required', 'matchRegexp:^[a-zA-Z0-9]{6,20}$']}
            errorMessages={['Please enter current password',
              'Password length must be between 6-20 characters and contains no special character']}
          />
          <TextValidator
            label='New Password'
            onChange={this.handleNewPasswordChange}
            name='newPassword'
            style={styles.textField}
            margin='dense'
            type='password'
            value={newPassword}
            helperText=' '
            validators={['required', 'matchRegexp:^[a-zA-Z0-9]{6,20}$']}
            errorMessages={['Please enter new password',
              'Password length must be between 6-20 characters and contains no special character']}
          />
          <TextValidator
            label='Confirm New Password'
            onChange={this.handleConfirmNewPasswordChange}
            name='confirmNewPassword'
            style={styles.textField}
            margin='dense'
            type='password'
            value={confirmNewPassword}
            helperText=' '
            validators={['required', 'matchRegexp:^[a-zA-Z0-9]{6,20}$', 'isNewPasswordMatch']}
            errorMessages={['Please confirm new password',
              'Password length must be between 6-20 characters and contains no special character',
              'Confirm new password must match new password']}
          />
          <Button
            variant='outlined'
            size='small'
            component='button'
            style={styles.button}
            disabled={this.disableChangePasswordButton()}
            type='submit'
          >
            {isChangingPassword ? <CircularProgress size={22} /> : accountString.changePassword}
          </Button>
        </ValidatorForm>
      </div>
    )
  }
}

export default connect(selector)(ChangePasswordForm)