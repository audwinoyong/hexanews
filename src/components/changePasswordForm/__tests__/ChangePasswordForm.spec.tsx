import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { shallow } from 'enzyme'

import { ChangePasswordForm, Props, ComponentState } from '../ChangePasswordForm'

jest.mock('react-quill')
Enzyme.configure({ adapter: new Adapter() })

const mockUser = {
  _id: '1',
  createdAt: 'created',
  description: 'desc',
  email: 'user@email.com',
  name: 'user'
}

describe('Change Password Form', () => {
  it('renders Change Password Form', () => {
    const component = renderer.create(
      <ChangePasswordForm
        user={mockUser}
        dispatch={jest.fn()}
        isChangingPassword={false}
        changePasswordError='test'
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Change Password with isChangingPassword as true', () => {
    const component = renderer.create(
      <ChangePasswordForm
        user={mockUser}
        dispatch={jest.fn()}
        isChangingPassword={true}
        changePasswordError='test'
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle handleCurrentPassword Change', () => {
    const component = (
      <ChangePasswordForm
        user={mockUser}
        dispatch={jest.fn()}
        isChangingPassword={true}
        changePasswordError='test'
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: ChangePasswordForm = layout.instance() as ChangePasswordForm

    const currentPassword = layout.state().currentPassword
    expect(currentPassword).toBe('')

    const event = {
      target: {
        value: 'a'
      }
    }

    instance.handleCurrentPasswordChange(event)
    layout.update()

    const updatedCurrentPassword = layout.state().currentPassword
    expect(updatedCurrentPassword).toBe('a')
  })

  it('handle handleNewPassword Change', () => {
    const component = (
      <ChangePasswordForm
        user={mockUser}
        dispatch={jest.fn()}
        isChangingPassword={true}
        changePasswordError='test'
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: ChangePasswordForm = layout.instance() as ChangePasswordForm

    const newPassword = layout.state().newPassword
    expect(newPassword).toBe('')

    const event = {
      target: {
        value: 'a'
      }
    }

    instance.handleNewPasswordChange(event)
    layout.update()

    const updatedNewPassword = layout.state().newPassword
    expect(updatedNewPassword).toBe('a')
  })

  it('handle handleConfirmNewPassword Change', () => {
    const component = (
      <ChangePasswordForm
        user={mockUser}
        dispatch={jest.fn()}
        isChangingPassword={true}
        changePasswordError='test'
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: ChangePasswordForm = layout.instance() as ChangePasswordForm

    const confirmNewPassword = layout.state().confirmNewPassword
    expect(confirmNewPassword).toBe('')

    const event = {
      target: {
        value: 'a'
      }
    }

    instance.handleConfirmNewPasswordChange(event)
    layout.update()

    const updatedConfirmNewPassword = layout.state().confirmNewPassword
    expect(updatedConfirmNewPassword).toBe('a')
  })

  it('handle onChangePassword action correctly', () => {
    const dispatch = jest.fn()
    const component = (
      <ChangePasswordForm
        user={mockUser}
        dispatch={dispatch}
        isChangingPassword={true}
        changePasswordError='test'
      />
    )

    const layout = shallow(component)
    const instance: ChangePasswordForm = layout.instance() as ChangePasswordForm

    instance.onChangePassword()
    layout.update()

    expect(dispatch).toHaveBeenCalled()
  })
})