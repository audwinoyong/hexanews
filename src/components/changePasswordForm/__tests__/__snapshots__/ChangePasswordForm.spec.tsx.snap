// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Change Password Form renders Change Password Form 1`] = `
<div
  style={
    Object {
      "marginTop": 24,
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="MuiTypography-root-1 MuiTypography-headline-6 MuiTypography-colorTextPrimary-23"
    >
      Change Password
    </h2>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginNormal-27"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34"
        data-shrink={false}
        htmlFor={undefined}
      >
        Current Password
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputType-53"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="currentPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginNormal-27"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34"
        data-shrink={false}
        htmlFor={undefined}
      >
        New Password
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputType-53"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="newPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginNormal-27"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34"
        data-shrink={false}
        htmlFor={undefined}
      >
        Confirm New Password
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputType-53"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="confirmNewPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-81 MuiButtonBase-disabled-82 MuiButton-root-55 MuiButton-text-57 MuiButton-outlined-63 MuiButton-sizeSmall-78 MuiButton-disabled-75"
      disabled={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginRight": 10,
          "marginTop": 30,
        }
      }
      tabIndex="-1"
      type="submit"
    >
      <span
        className="MuiButton-label-56"
      >
        Change Password
      </span>
    </button>
  </form>
</div>
`;

exports[`Change Password Form renders Change Password with isChangingPassword as true 1`] = `
<div
  style={
    Object {
      "marginTop": 24,
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="MuiTypography-root-1 MuiTypography-headline-6 MuiTypography-colorTextPrimary-23"
    >
      Change Password
    </h2>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginNormal-27"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34"
        data-shrink={false}
        htmlFor={undefined}
      >
        Current Password
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputType-53"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="currentPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginNormal-27"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34"
        data-shrink={false}
        htmlFor={undefined}
      >
        New Password
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputType-53"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="newPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginNormal-27"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34"
        data-shrink={false}
        htmlFor={undefined}
      >
        Confirm New Password
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputType-53"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="confirmNewPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-81 MuiButtonBase-disabled-82 MuiButton-root-55 MuiButton-text-57 MuiButton-outlined-63 MuiButton-sizeSmall-78 MuiButton-disabled-75"
      disabled={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginRight": 10,
          "marginTop": 30,
        }
      }
      tabIndex="-1"
      type="submit"
    >
      <span
        className="MuiButton-label-56"
      >
        <div
          className="MuiCircularProgress-root-84 MuiCircularProgress-colorPrimary-87 MuiCircularProgress-indeterminate-86"
          role="progressbar"
          style={
            Object {
              "height": 22,
              "width": 22,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg-89"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle-90 MuiCircularProgress-circleIndeterminate-92"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </div>
      </span>
    </button>
  </form>
</div>
`;
